service/backend/pchain/indexer/parser.go

```diff
@@ -0,0 +1,329 @@
+}
+
+func NewParser(pChainClient client.PChainClient) (Parser, error) {
+	errs := wrappers.Errs{}
+
+	aliaser := ids.NewAliaser()
+	errs.Add(aliaser.Alias(constants.PlatformChainID, mapper.PChainNetworkIdentifier))
+
+	return &parser{
+		codec:            blocks.Codec,
+		codecVersion:     txs.Version,
+		pChainClient:     pChainClient,
+		aliaser:          aliaser,
+		genesisTimestamp: time.Unix(genesisTimestamp, 0),
+	}, errs.Err
+}
+
+func (p *parser) initCtx(ctx context.Context) error {
+	if p.ctx == nil {
+		networkID, err := p.pChainClient.GetNetworkID(ctx)
+		if err != nil {
+			return err
+		}
+
+		p.networkID = networkID
+		p.ctx = &snow.Context{
+			BCLookup:  p.aliaser,
+			NetworkID: networkID,
+		}
+	}
+
+	return nil
+}
+
+func (p *parser) GetPlatformHeight(ctx context.Context) (uint64, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	return p.pChainClient.GetHeight(ctx)
+}
+
+func (p *parser) GetGenesisBlock(ctx context.Context) (*ParsedGenesisBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	errs := wrappers.Errs{}
+
+	bytes, avaxAssetID, err := genesis.FromConfig(genesis.GetConfig(p.networkID))
+	errs.Add(err)
+	p.avaxAssetID = avaxAssetID
+
+	genesisState, err := pGenesis.Parse(bytes)
+	errs.Add(err)
+
+	p.genesisTimestamp = time.Unix(int64(genesisState.Timestamp), 0)
+
+	var genesisTxs []*txs.Tx
+	genesisTxs = append(genesisTxs, genesisState.Validators...)
+	genesisTxs = append(genesisTxs, genesisState.Chains...)
+
+	// Genesis commit block's parent ID is the hash of genesis state
+	var genesisParentID ids.ID = hashing.ComputeHash256Array(bytes)
+
+	// Genesis Block is not indexed by the indexer, but its block ID can be accessed from block 0's parent id
+	genesisChildBlock, err := p.ParseBlockAtIndex(ctx, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, utxo := range genesisState.UTXOs {
+		utxo.UTXO.Out.InitCtx(p.ctx)
+	}
+
+	genesisBlockID := genesisChildBlock.ParentID
+
+	return &ParsedGenesisBlock{
+		ParsedBlock: ParsedBlock{
+			ParentID:  genesisParentID,
+			Height:    0,
+			BlockID:   genesisBlockID,
+			BlockType: "GenesisBlock",
+			Timestamp: p.genesisTimestamp.UnixMilli(),
+			Txs:       genesisTxs,
+			Proposer:  Proposer{},
+		},
+		GenesisBlockData: GenesisBlockData{
+			Message:       genesisState.Message,
+			InitialSupply: genesisState.InitialSupply,
+			UTXOs:         genesisState.UTXOs,
+		},
+	}, errs.Err
+}
+
+func (p *parser) ParseCurrentBlock(ctx context.Context) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	height, err := p.GetPlatformHeight(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.ParseBlockAtIndex(ctx, height)
+}
+
+func (p *parser) ParseBlockAtIndex(ctx context.Context, index uint64) (*ParsedBlock, error) {
+	err := p.initCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByIndex(ctx, index-1)
```

Comment:
I'm not sure if I understand what you are asking. Container indices seem to be off by one compared to block heights in P-chain (e.g. block 10 seems to be indexed in container id 9) and hence we have special handling for the genesis block. I need to perform the -1 somewhere, are you suggesting we should do it in the clients of this method instead of here?

Reply:
Let me cross check on avalanche go side. May be a behaviour I missed (maybe genesis is not indexed)

Reply:
You are right. So index API does not index genesis block. So height n maps to index n-1.
Thanks for bringing it up